mathoxide-lib/src/views.rs

```diff
@@ -1,47 +1,90 @@
+use crate::view_iters::ContiguousViewIterator;
+
 pub trait ArrayView {
+    type IterType;
+
     fn translate<ListType: AsRef<[usize]>>(&self, idx: ListType) -> usize;
     fn offset(&self) -> usize;
-    fn ndims(&self) -> usize;
-    fn size(&self) -> usize;
     fn shape(&self) -> &[usize];
+    fn stride(&self) -> &[usize];
+    fn ndim(&self) -> usize {
+        self.shape().len()
+    }
+    fn numel(&self) -> usize {
+        self.shape().iter().product()
+    }
+    fn is_contiguous(&self) -> bool;
+    fn translate_iter(&self) -> Self::IterType;
 }
 
-pub struct SimpleView {
+pub struct ContiguousView {
     shape: Vec<usize>,
+    offset: usize,
+    stride: Vec<usize>,
 }
 
-impl SimpleView {
+impl ContiguousView {
     pub fn new<ListType: AsRef<[usize]>>(shape: ListType) -> Self {
+        Self::new_with_offset(shape, 0)
+    }
+
+    pub fn new_with_offset<ListType: AsRef<[usize]>>(shape: ListType, offset: usize) -> Self {
         Self {
             shape: shape.as_ref().to_vec(),
+            offset,
+            stride: Self::compute_stride(shape.as_ref()),
         }
     }
+
+    fn compute_stride<ListType: AsRef<[usize]>>(shape: ListType) -> Vec<usize> {
+        let mut res = shape
+            .as_ref()
+            .iter()
+            .rev()
+            .scan(1, |state, &x| {
+                let tmp = *state;
+                *state *= x;
+                Some(tmp)
+            })
+            .collect::<Vec<usize>>();
+        res.reverse();
+        res
+    }
 }
 
-impl ArrayView for SimpleView {
+impl ArrayView for ContiguousView {
+    type IterType = ContiguousViewIterator;
+
     fn translate<ListType: AsRef<[usize]>>(&self, idx: ListType) -> usize {
-        let mut res: usize = 0;
-        for (idx_i, shape_i) in idx.as_ref().iter().zip(self.shape.iter()) {
-            res *= shape_i;
-            res += idx_i;
-        }
-        res
+        self.offset()
+            + idx
+                .as_ref()
+                .iter()
+                .zip(self.stride().iter())
+                .map(|(x, y)| x * y)
+                .sum::<usize>()
     }
 
     fn offset(&self) -> usize {
-        0
+        self.offset
     }
 
-    fn ndims(&self) -> usize {
-        self.shape.len()
+    fn shape(&self) -> &[usize] {
+        self.shape.as_slice()
     }
 
-    fn size(&self) -> usize {
-        self.shape.iter().product()
+    fn stride(&self) -> &[usize] {
+        self.stride.as_slice()
     }
 
-    fn shape(&self) -> &[usize] {
-        self.shape.as_slice()
+    fn is_contiguous(&self) -> bool {
+        true
+    }
+
+    fn translate_iter(&self) -> Self::IterType {
+        let begin = self.translate(vec![0; self.ndim()]);
```

Comment:
this is effectively `self.offset()`. I am not sure if it is better to use that directly or to keep the code as is. Using the `offset` method is more explicit, and I don't think the `ContiguousView` will ever need to not return `self.offset()` as the first item.